plugins/outputs/iotdb/iotdb.go

```diff
@@ -0,0 +1,287 @@
+			return client.INT64, int64(v)
+		} else if s.ConvertUint64To == "Text" {
+			return client.TEXT, fmt.Sprintf("%d", v)
+		} else {
+			s.Log.Errorf("unknown converstaion configuration of 'convertUint64To': %s", s.ConvertUint64To)
+			return client.UNKNOW, int64(0)
+		}
+	case float64:
+		return client.DOUBLE, v
+	case string:
+		return client.TEXT, v
+	case bool:
+		return client.BOOLEAN, v
+	default:
+		s.Log.Errorf("Unknown datatype: '%T' %v", value, value)
+		return client.UNKNOW, int64(0)
+	}
+}
+
+// convert Timestamp Unit accroding to config
+func (s *IoTDB) ConvertTimestampOfMetric(m telegraf.Metric) (int64, error) {
+	if s.TimeStampUnit == "second" {
+		return m.Time().Unix(), nil
+	} else if s.TimeStampUnit == "millisecond" {
+		return m.Time().UnixMilli(), nil
+	} else if s.TimeStampUnit == "microsecond" {
+		return m.Time().UnixMicro(), nil
+	} else if s.TimeStampUnit == "nanosecond" {
+		return m.Time().UnixNano(), nil
+	}
+	errorMsg := fmt.Sprintf("IoTDB Configuration Error: unknown TimeStampUnit: %s", s.TimeStampUnit)
+	s.Log.Errorf(errorMsg)
+	return 0, errors.New(errorMsg)
+}
+
+// convert Metrics to Records with tags
+func (s *IoTDB) ConvertMetricsToRecordsWithTags(metrics []telegraf.Metric) (*RecordsWithTags, error) {
+	var deviceidList []string
+	var measurementsList [][]string
+	var valuesList [][]interface{}
+	var dataTypesList [][]client.TSDataType
+	var timestampList []int64
+	var tagsList [][]*telegraf.Tag
+
+	for _, metric := range metrics {
+		// write `metric` to the output sink here
+		var tags []*telegraf.Tag
+		tags = append(tags, metric.TagList()...)
+		// deal with basic paramter
+		var keys []string
+		var values []interface{}
+		var dataTypes []client.TSDataType
+		for _, field := range metric.FieldList() {
+			datatype, value := s.getDataTypeAndValue(field.Value)
+			if datatype != client.UNKNOW {
+				keys = append(keys, field.Key)
+				values = append(values, value)
+				dataTypes = append(dataTypes, datatype)
+			}
+		}
+		// Convert timestamp into specified unit
+		ts, tsError := s.ConvertTimestampOfMetric(metric)
+		if tsError != nil {
+			return nil, tsError
+		}
+		timestampList = append(timestampList, ts)
+		// append all metric data of this record to lists
+		deviceidList = append(deviceidList, metric.Name())
+		measurementsList = append(measurementsList, keys)
+		valuesList = append(valuesList, values)
+		dataTypesList = append(dataTypesList, dataTypes)
+		tagsList = append(tagsList, tags)
+	}
+	var recordsWithTags = &RecordsWithTags{
+		DeviceIDList:     deviceidList,
+		MeasurementsList: measurementsList,
+		ValuesList:       valuesList,
+		DataTypesList:    dataTypesList,
+		TimestampList:    timestampList,
+		TagsList:         tagsList,
+	}
+	return recordsWithTags, nil
+}
+
+// modifiy RecordsWithTags according to 'TreateTagsAs' Configuration
+func (s *IoTDB) ModifiyRecordsWithTags(rwt *RecordsWithTags) error {
+	if s.TreateTagsAs == "Measurements" {
```

Comment:
Do not export the function. Use a switch statement here. Shorten your error handling.

Reply:
Tanks for reminding me! I will fix them.